scarlet/renderer.py

```diff
@@ -0,0 +1,520 @@
+            array for linear mapping of model channels onto observation channels
+        """
+
+        if list(data_frame.channels) == list(model_frame.channels):
+            return None
+
+        channel_map = [
+            list(model_frame.channels).index(c) for c in list(data_frame.channels)
+        ]
+        min_channel = min(channel_map)
+        max_channel = max(channel_map)
+        if max_channel + 1 - min_channel == len(channel_map):
+            channel_map = slice(min_channel, max_channel + 1)
+        return channel_map
+
+        # full-fledged linear mixing model to allow for spectrophotometry later
+        channel_map = np.zeros((data_frame.C, model_frame.C))
+        for i, c in enumerate(list(data_frame.channels)):
+            j = list(model_frame.channels).index(c)
+            assert j != -1, f"Could not find channel {c} in model frame"
+            channel_map[i, j] = 1
+            # TODO: for overlap computation:
+            # * turn channels into dict channel['g'] = (lambdas, R_lambdas)
+            # * extrapolate obs R_lambda onto model lambdas
+            # * compute np.dot(model_R_lambda, obs_R_lambda) for every
+            #   combination of obs and model channels
+        return channel_map
+
+    def map_channels(self, model):
+        """Map to model channels onto the observation channels
+
+        Parameters
+        ----------
+        model: array
+            The hyperspectral model
+
+        Returns
+        -------
+        obs_model: array
+            `model` mapped onto the observation channels
+        """
+        if self.channel_map is None:
+            return model
+        if isinstance(self.channel_map, slice):
+            return model[self.channel_map]
+        return np.dot(self.channel_map, model)
+
+
+class NullRenderer(Renderer):
+    def __init__(self):
+        super().__init__(None, None)
+
+    def __call__(self, model, *parameters):
+        return model
+
+
+@primitive
+def convolve(image, psf, bounds):
+    """Convolve an image with a PSF in real space
+    """
+    result = np.empty(image.shape, dtype=image.dtype)
+    for band in range(len(image)):
+        if hasattr(image[band], "_value"):
+            # This is an ArrayBox
+            img = image[band]._value
+        else:
+            img = image[band]
+        apply_filter(
+            img,
+            psf[band].reshape(-1),
+            bounds[0],
+            bounds[1],
+            bounds[2],
+            bounds[3],
+            result[band],
+        )
+    return result
+
+
+def _grad_convolve(convolved, image, psf, slices):
+    """Gradient of a real space convolution
+    """
+    return lambda input_grad: convolve(input_grad, psf[:, ::-1, ::-1], slices)
+
+
+# Register this function in autograd
+defvjp(convolve, _grad_convolve)
+
+# match the spatial shapes of model and data
+@primitive
+def match_shape(model, data_frame, slices):
+    data_slices, model_slices = slices
```

Comment:
I'm surprised that you prefer this method to just calculating the slices once using overlapped_slices and caching that. Can you explain why you want to add something seemingly unecessary to the gradient path instead?

Reply:
The logic is the following: The renderer should deliver the model rendered on the whole observation frame. Observation actually doesn't know anything about slices, channels, etc. Thus, we have to match the shapes. And you will recognize the code for doing that...